File: sdk/npm/src/providers/tempmail-lol.ts
import { EmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';

const CHANNEL: Channel = 'tempmail-lol';
const BASE_URL = 'https://api.tempmail.lol/v2';

const DEFAULT_HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/143.0.0.0 Safari/537.36',
  'Content-Type': 'application/json',
  'Origin': 'https://tempmail.lol',
  'sec-ch-ua': '"Microsoft Edge";v="143", "Chromium";v="143", "Not A(Brand";v="24"',
  'sec-ch-ua-mobile': '?0',
  'sec-ch-ua-platform': '"Windows"',
  'DNT': '1',
};

export async function generateEmail(domain: string | null = null): Promise<EmailInfo> {
  const response = await fetch(`${BASE_URL}/inbox/create`, {
    method: 'POST',
    headers: DEFAULT_HEADERS,
    body: JSON.stringify({ domain, captcha: null }),
  });

  if (!response.ok) {
    throw new Error(`Failed to generate email: ${response.status}`);
  }

  const data = await response.json();
  
  if (!data.address || !data.token) {
    throw new Error('Failed to generate email');
  }

  return {
    channel: CHANNEL,
    email: data.address,
    token: data.token,
  };
}

export async function getEmails(token: string, recipientEmail: string = ''): Promise<Email[]> {
  const response = await fetch(`${BASE_URL}/inbox?token=${encodeURIComponent(token)}`, {
    method: 'GET',
    headers: DEFAULT_HEADERS,
  });

  if (!response.ok) {
    throw new Error(`Failed to get emails: ${response.status}`);
  }

  const data = await response.json();
  
  const rawEmails = data.emails || [];
  return rawEmails.map((raw: any) => normalizeEmail(raw, recipientEmail));
}

File: sdk/go/provider/tempmail_lol.go
package provider

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/url"

	http "github.com/bogdanfinn/fhttp"
)

const tempmailLolBaseURL = "https://api.tempmail.lol/v2"

type TempmailLolGenerateRequest struct {
	Domain  *string `json:"domain"`
	Captcha *string `json:"captcha"`
}

type TempmailLolGenerateResponse struct {
	Address string `json:"address"`
	Token   string `json:"token"`
}

type tempmailLolEmailsResponse struct {
	Emails  []json.RawMessage `json:"emails"`
	Expired bool              `json:"expired"`
}

func TempmailLolGenerate(domain *string) (*CreatedMailbox, error) {
	reqBody, _ := json.Marshal(TempmailLolGenerateRequest{Domain: domain, Captcha: nil})

	req, err := http.NewRequest("POST", tempmailLolBaseURL+"/inbox/create", bytes.NewReader(reqBody))
	if err != nil {
		return nil, err
	}

	req.Header.Set("User-Agent", GetCurrentUA())
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Origin", "https://tempmail.lol")
	req.Header.Set("DNT", "1")

	client := HTTPClient()
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if err := CheckHTTPStatus(resp, "tempmail-lol generate"); err != nil {
		return nil, err
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var result TempmailLolGenerateResponse
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}

	if result.Address == "" || result.Token == "" {
		return nil, fmt.Errorf("failed to generate email")
	}

	info := &CreatedMailbox{Channel: "tempmail-lol", Email: result.Address, Token: result.Token}
	return info, nil
}

func TempmailLolGetEmails(token string, recipientEmail string) ([]NormEmail, error) {
	encodedToken := url.QueryEscape(token)

	req, err := http.NewRequest("GET", tempmailLolBaseURL+"/inbox?token="+encodedToken, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("User-Agent", GetCurrentUA())
	req.Header.Set("Origin", "https://tempmail.lol")
	req.Header.Set("DNT", "1")

	client := HTTPClient()
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if err := CheckHTTPStatus(resp, "tempmail-lol get emails"); err != nil {
		return nil, err
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var result tempmailLolEmailsResponse
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}

	return NormalizeRawMessages(result.Emails, recipientEmail)
}
